frontend/src/components/ui/file-input.ts

```diff
@@ -0,0 +1,240 @@
+  /**
+   * Form control name, if used as a form control
+   */
+  @property({ type: String })
+  name?: string;
+
+  /**
+   * Form control label, if used as a form control
+   */
+  @property({ type: String })
+  label?: string;
+
+  /**
+   * Specify which file types are allowed
+   */
+  @property({ type: String })
+  accept?: HTMLInputElement["accept"];
+
+  /**
+   * Enable selecting more than one file
+   */
+  @property({ type: Boolean })
+  multiple?: HTMLInputElement["multiple"];
+
+  /**
+   * Enable dragging files into drop zone
+   */
+  @property({ type: Boolean })
+  drop = false;
+
+  @state()
+  private files: File[] = [];
+
+  @query("#dropzone")
+  private readonly dropzone?: HTMLElement | null;
+
+  @query("input[type='file']")
+  private readonly input?: HTMLInputElement | null;
+
+  formResetCallback() {
+    this.files = [];
+
+    if (this.input) {
+      this.input.value = "";
+    }
+  }
+
+  protected willUpdate(changedProperties: PropertyValues): void {
+    if (changedProperties.has("files")) {
+      this.syncFormValue();
+    }
+  }
+
+  private syncFormValue() {
+    const formControlName = this.name;
+
+    if (!formControlName) return;
+
+    // `ElementInternals["setFormValue"]` doesn't support `FileList` yet,
+    // construct `FormData` instead
+    const formData = new FormData();
+
+    this.files.forEach((file) => {
+      formData.append(formControlName, file);
+    });
+
+    this.setFormValue(formData);
+  }
+
+  render() {
+    return html`
+      ${this.label
+        ? html`<label for="fileInput" class="form-label">${this.label}</label>`
+        : nothing}
+      ${this.files.length ? this.renderFiles() : this.renderInput()}
+    `;
+  }
+
+  private readonly renderInput = () => {
+    return html`
+      <div
+        id="dropzone"
+        class=${clsx(
+          tw`relative`,
+          this.drop &&
+            tw`flex size-full items-center justify-center rounded p-6 text-center outline-dashed outline-1 -outline-offset-1 outline-neutral-400 transition-all hover:outline-primary-400`,
+        )}
+        @drop=${this.drop ? this.onDrop : undefined}
+        @dragover=${this.drop ? this.onDragover : undefined}
+        @dragenter=${this.drop
+          ? () => this.dropzone?.classList.add(droppingClass)
+          : undefined}
+        @dragleave=${this.drop
+          ? () => this.dropzone?.classList.remove(droppingClass)
+          : undefined}
+        dropzone="copy"
+        aria-dropeffect="copy"
+      >
+        <input
+          id="fileInput"
+          class="sr-only"
+          type="file"
+          accept=${ifDefined(this.accept)}
+          ?multiple=${this.multiple}
+          @change=${() => {
+            const files = this.input?.files;
+
+            if (files) {
+              void this.handleChange(files);
+            }
+          }}
+        />
+        <div class="relative z-10">
+          <slot @click=${() => this.input?.click()}></slot>
```

Comment:
Do we want to move this to the `#dropzone` div so that the whole area is clickable?